File: 01-overview/homework_quest/static/homework_quest/offline_queue.js
/**
 * Offline queue for log/approve POSTs (localStorage).
 * See homework_quest/offline_queue.py for payload schema.
 */
(function (global) {
  var STORAGE_KEY = "homework_quest_offline_queue_v1";

  function loadQueue() {
    try {
      var raw = global.localStorage.getItem(STORAGE_KEY);
      if (!raw) return { version: 1, items: [] };
      var parsed = JSON.parse(raw);
      if (!parsed.items) parsed.items = [];
      parsed.version = 1;
      return parsed;
    } catch (err) {
      return { version: 1, items: [] };
    }
  }

  function saveQueue(queue) {
    global.localStorage.setItem(STORAGE_KEY, JSON.stringify(queue));
  }

  function enqueue(item) {
    var queue = loadQueue();
    var entry = {
      id: String(Date.now()) + "-" + String(Math.random()).slice(2, 8),
      kind: item.kind,
      url: item.url,
      method: "POST",
      contentType: item.contentType,
      body: item.body,
      createdAt: new Date().toISOString(),
    };
    queue.items.push(entry);
    saveQueue(queue);
    return entry;
  }

  function sendItem(item) {
    if (item.contentType === "json") {
      return fetch(item.url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(item.body),
      });
    }
    var formData = new FormData();
    Object.keys(item.body).forEach(function (key) {
      formData.append(key, item.body[key]);
    });
    return fetch(item.url, {
      method: "POST",
      body: formData,
      credentials: "same-origin",
    });
  }

  function flush() {
    if (!global.navigator.onLine) {
      return Promise.resolve({ sent: 0, remaining: loadQueue().items.length });
    }
    var queue = loadQueue();
    var remaining = [];
    var sent = 0;
    var chain = Promise.resolve();

    queue.items.forEach(function (item) {
      chain = chain.then(function () {
        return sendItem(item).then(function (response) {
          if (response.ok || response.redirected || response.status === 302) {
            sent += 1;
          } else {
            remaining.push(item);
          }
        }).catch(function () {
          remaining.push(item);
        });
      });
    });

    return chain.then(function () {
      saveQueue({ version: 1, items: remaining });
      return { sent: sent, remaining: remaining.length };
    });
  }

  function queueOrSend(request, onSuccess, onQueued, onError) {
    if (!global.navigator.onLine) {
      enqueue(request);
      if (onQueued) onQueued();
      return Promise.resolve(false);
    }
    return sendItem(request)
      .then(function (response) {
        if (response.ok || response.redirected || response.status === 302) {
          if (onSuccess) onSuccess(response);
          return true;
        }
        enqueue(request);
        if (onQueued) onQueued();
        return false;
      })
      .catch(function () {
        enqueue(request);
        if (onQueued) onQueued();
        return false;
      });
  }

  global.HomeworkQuestOfflineQueue = {
    STORAGE_KEY: STORAGE_KEY,
    loadQueue: loadQueue,
    enqueue: enqueue,
    flush: flush,
    queueOrSend: queueOrSend,
  };

  global.addEventListener("online", function () {
    flush();
  });

  document.addEventListener("DOMContentLoaded", function () {
    flush();
  });
})(window);

File: 01-overview/homework_quest/offline_queue.py
"""
Offline queue payload validation (mirrors homework_quest/static/homework_quest/offline_queue.js).

Payload shape stored in localStorage under key ``homework_quest_offline_queue_v1``:

{
  "version": 1,
  "items": [
    {
      "id": "<unique string>",
      "kind": "approve" | "log-form",
      "url": "<POST url>",
      "method": "POST",
      "contentType": "json" | "form",
      "body": { ... },
      "createdAt": "<ISO-8601>"
    }
  ]
}

``approve`` body: {"approver_id": int, "pin": "1234"}
``log-form`` body: {"profile_id": str, "pin": "1234", "csrfmiddlewaretoken": str}
"""

from __future__ import annotations

VALID_KINDS = frozenset({"approve", "log-form"})
VALID_CONTENT_TYPES = frozenset({"json", "form"})


class OfflineQueueError(Exception):
    pass


def validate_queue_item(item: dict) -> dict:
    if not isinstance(item, dict):
        raise OfflineQueueError("Queue item must be an object.")
    kind = item.get("kind")
    if kind not in VALID_KINDS:
        raise OfflineQueueError(f"Unsupported kind: {kind!r}")
    if not item.get("url"):
        raise OfflineQueueError("Queue item requires url.")
    content_type = item.get("contentType")
    if content_type not in VALID_CONTENT_TYPES:
        raise OfflineQueueError(f"Unsupported contentType: {content_type!r}")
    body = item.get("body")
    if not isinstance(body, dict):
        raise OfflineQueueError("Queue item body must be an object.")
    if kind == "approve":
        if "approver_id" not in body or "pin" not in body:
            raise OfflineQueueError("Approve payload requires approver_id and pin.")
    if kind == "log-form":
        for key in ("profile_id", "pin"):
            if key not in body:
                raise OfflineQueueError(f"log-form payload requires {key}.")
    return item


def validate_queue_document(document: dict) -> dict:
    if document.get("version") != 1:
        raise OfflineQueueError("Unsupported queue version.")
    items = document.get("items")
    if not isinstance(items, list):
        raise OfflineQueueError("Queue items must be a list.")
    for item in items:
        validate_queue_item(item)
    return document
